server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -3354,6 +3354,11 @@ private <T> void executeShardSnapshotUpdate(
                     updatedState = updateSnapshotState.updatedState;
                 }
 
+                if (updatedState.state() == ShardState.PAUSED_FOR_NODE_REMOVAL) {
+                    // leave subsequent entries for this shard alone until this one is unpaused
+                    iterator.remove();
+                }
```

Comment:
It still feels a bit odd that `tryStartNextTaskAfterSnapshotUpdated` can init shard snapshot without  checking whether this shard is already active in snapshots that come before it. I think it relied on the fact that the possible updated states sent by data nodes are all "non-active" which is true till we add `PAUSED_FOR_NODE_REMOVAL`. Theoretically this could happen again in future if we add another similar ShardState. That said, I can see it would be quite a lot work to change things in `tryStartNextTaskAfterSnapshotUpdated`. So I am good with this fix.

Reply:
++ valid point, I think we can at least add an assertion to catch a future bug in this area - see 615cebf49ea.